gapil/runtime/cc/runtime.h

```diff
@@ -102,18 +104,81 @@ typedef struct ref_t {
 // buffer is a structure used to hold a variable size byte array.
 // buffer is used internally by the compiler to write out variable length data.
 typedef struct buffer_t {
-  uint8_t* data;      // buffer data.
-  uint32_t capacity;  // total capacity of the buffer.
-  uint32_t size;      // current size of the buffer.
+  arena* arena;        // the arena that owns the buffer data.
+  uint8_t* data;       // buffer data.
+  uint32_t capacity;   // total capacity of the buffer.
+  uint32_t size;       // current size of the buffer.
+  uint32_t alignment;  // min alignment in bytes of the data allocation.
 } buffer;
 
+#define GAPIL_KIND_BOOL 1
+#define GAPIL_KIND_U8 2
+#define GAPIL_KIND_S8 3
+#define GAPIL_KIND_U16 4
+#define GAPIL_KIND_S16 5
+#define GAPIL_KIND_F32 6
+#define GAPIL_KIND_U32 7
+#define GAPIL_KIND_S32 8
+#define GAPIL_KIND_F64 9
+#define GAPIL_KIND_U64 10
+#define GAPIL_KIND_S64 11
+#define GAPIL_KIND_INT 12
+#define GAPIL_KIND_UINT 13
+#define GAPIL_KIND_SIZE 14
+#define GAPIL_KIND_CHAR 15
+#define GAPIL_KIND_ARRAY 16
+#define GAPIL_KIND_CLASS 17
+#define GAPIL_KIND_ENUM 18
+#define GAPIL_KIND_MAP 19
+#define GAPIL_KIND_POINTER 20
+#define GAPIL_KIND_REFERENCE 21
+#define GAPIL_KIND_SLICE 22
+#define GAPIL_KIND_STRING 23
+
+typedef struct gapil_rtti {
+  uint32_t kind;             // kind of the type.
+  uint32_t api_index;        // api index to which the type belongs.
+  uint32_t type_index;       // index of the type within the api.
+  gapil_char* type_name;     // name of the type.
+  void (*reference)(void*);  // increment the reference count for the type.
+  void (*release)(void*);    // decrement the reference count for the type.
+} gapil_rtti;
+
+typedef struct gapil_any_t {
+  uint32_t ref_count;  // number of owners of this any.
+  arena* arena;        // arena that owns this any allocation.
+  gapil_rtti* rtti;    // type information of the value.
+  void* value;  // pointer to the value. For boxed value-types, this should be
```

Comment:
Just for my own sake, does this mean any<float> would "effectively" be
gapil_any_t {
     uint32_t ref_count;
     arena* arena;
     gapil_rtti* rtti;
     void* value;   (= &hidden_value);
     float _hidden_value;
}

Reply:
Yes. I considered unions, but these are a nightmare to deal with in LLVM.
Given that they are infrequently used (both in the API and in typical numbers created), I've decided to leave this as-is for now. If the double indirection ever shows up in a profile, we can optimize.